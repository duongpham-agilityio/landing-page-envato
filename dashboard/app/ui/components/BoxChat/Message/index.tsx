import { Avatar, Box, Flex, Spacer, Text } from '@chakra-ui/react';
import { useEffect, useState } from 'react';

// Constants
import { IMAGES } from '@/lib/constants';

// Hooks
import { getUsers } from '@/lib/hooks';

interface MessageProps {
  content?: string;
  avatarUser?: string;
  avatarAdmin?: string;
  localeTime?: string;
  senderId: string;
}

const Message = ({
  senderId,
  content = '',
  avatarUser = IMAGES.CHAT_USER_AVATAR.url,
  avatarAdmin = IMAGES.CHAT_USER_AVATAR.url,
  localeTime = new Date().toLocaleTimeString([], {
    hour: 'numeric',
    minute: '2-digit',
  }),
}: MessageProps) => {
  // TODO: instead of USER_ID use real id
  // const userId = authStore((state) => state.user?.id);
  const [isAdmin, setIsAdmin] = useState<string>('');

  useEffect(() => {
    const fetchData = async () => {
      const usersData = await getUsers();
      setIsAdmin(usersData?.adminId || '');
    };

    fetchData();
  }, []);

  return (
    <Flex
      width="100%"
      justifyContent={senderId !== isAdmin ? 'flex-end' : 'flex-start'}
    >
      {senderId === isAdmin && (
        <Avatar
          src={avatarAdmin}
          borderColor="border.tertiary"
          w={9}
          h={9}
          mr={2}
        />
      )}

      <Flex
        align="flex-end"
        direction={senderId !== isAdmin ? 'row-reverse' : 'row'}
        mb="30px"
        alignItems="center"
      >
        <Box
          data-testid="image-container"
          bg={
            senderId !== isAdmin
              ? 'primary.300'
              : 'background.section.messageUser'
          }
          p={3}
          ml={2}
          borderRadius={8}
          color="text.primary"
          dangerouslySetInnerHTML={{ __html: content ?? '' }}
        />
        <Spacer />

        <Text
          ml={3}
          mr={3}
          fontSize="xs"
          color="text.textTime"
          fontWeight="medium"
          minW="max-content"
        >
          {localeTime}
        </Text>
      </Flex>

      {senderId !== isAdmin && (
        <Avatar
          src={avatarUser}
          borderColor="border.tertiary"
          w={9}
          h={9}
          ml={2}
          data-testid="avatar"
        />
      )}
    </Flex>
  );
};
export default Message;
