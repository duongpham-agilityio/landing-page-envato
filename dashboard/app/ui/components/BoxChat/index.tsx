'use client';

import { memo, useCallback, useEffect, useRef, useState } from 'react';
import { Box, Heading, Flex } from '@chakra-ui/react';

// Components
import Message from './Message';

// Components
import { InputSendMessages } from '@/ui/components';

// Interface
import { TMessages } from '@/lib/interfaces';

// Stores
import { authStore } from '@/lib/stores';

// Hooks
import { getCurrentUser } from '@/lib/hooks';

// Firebase
import {
  doc,
  getDoc,
  onSnapshot,
  serverTimestamp,
  setDoc,
  updateDoc,
} from 'firebase/firestore';
import { db } from '@/lib/utils';
import { FIREBASE_CHAT, USER_CHATS_FIELD } from '@/lib/constants';

const initialUserChat = {
  roomChatId: '',
  userId: '',
  adminId: '',
  avatarUrl: '',
  avatarAdminUrl: '',
  displayName: '',
};

const BoxChatComponent = () => {
  const [messages, setMessages] = useState<TMessages[]>([]);
  const user = authStore((state) => state.user);
  const [userChat, setUserChat] = useState(initialUserChat);
  const boxRef = useRef<HTMLDivElement | null>(null);

  const createChatRoom = useCallback(async () => {
    // Get user data
    const usersData = await getCurrentUser(user);

    if (usersData) {
      await setDoc(doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId), {
        messages: [],
      });

      await updateDoc(doc(db, FIREBASE_CHAT.USER_CHATS, usersData.adminId), {
        [usersData.roomChatId + USER_CHATS_FIELD.USER_INFO]: {
          uid: usersData.adminId,
          displayName: usersData.displayName,
          photoURL: usersData.avatarUrl,
        },
        [usersData.roomChatId + USER_CHATS_FIELD.DATE]: serverTimestamp(),
      });

      await updateDoc(doc(db, FIREBASE_CHAT.USER_CHATS, usersData.userId), {
        [usersData.roomChatId + USER_CHATS_FIELD.USER_INFO]: {
          uid: usersData.userId,
          displayName: usersData.displayName,
          photoURL: usersData.avatarUrl,
        },
        [usersData.roomChatId + USER_CHATS_FIELD.DATE]: serverTimestamp(),
      });
    }
  }, [user]);

  const fetchData = async () => {
    // Get user data
    const usersData = await getCurrentUser(user);

    // Check if usersData is undefined before accessing its properties
    if (usersData) {
      const res = await getDoc(
        doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId),
      );

      setUserChat(usersData);
      if (res.exists()) {
        setMessages(res.data().messages);
      } else {
        await createChatRoom();
      }
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  useEffect(() => {
    if (!userChat.roomChatId) return;
    const unSub = onSnapshot(
      doc(db, FIREBASE_CHAT.CHATS, userChat.roomChatId),
      (doc) => {
        doc.exists() && setMessages(doc.data().messages);
      },
    );

    if (boxRef.current) {
      boxRef.current.scrollTop = boxRef.current.scrollHeight;
    }

    return () => {
      unSub();
    };
  }, [userChat.roomChatId]);

  return (
    <Box w="full" bg="background.body.quaternary" borderRadius="lg">
      <Flex
        direction="row"
        justifyContent="space-between"
        alignItems="center"
        justify="center"
        borderBottom="1px solid"
        borderBottomColor="border.tertiary"
        padding="24px 26px"
      >
        <Heading
          as="h3"
          fontWeight="semibold"
          color="text.primary"
          fontSize="2xl"
          textTransform="capitalize"
        >
          team chat
        </Heading>
      </Flex>

      <Box padding={{ base: '24px 20px', lg: '45px 35px' }}>
        <Box
          ref={boxRef}
          overflowX="auto"
          overflowY="scroll"
          css={{
            '&::-webkit-scrollbar': {
              width: 2,
            },
            '&::-webkit-scrollbar-track': {
              width: 2,
            },
            '&::-webkit-scrollbar-thumb': {
              background: 'gray',
              borderRadius: '24px',
            },
          }}
          maxHeight={361}
          padding={5}
        >
          {messages.map((m) => (
            <Message
              content={m.text}
              key={m.date.second}
              senderId={m.senderId}
              avatarAdmin={userChat.avatarAdminUrl}
              avatarUser={userChat.avatarUrl}
            />
          ))}
        </Box>
        <InputSendMessages boxRef={boxRef} />
      </Box>
    </Box>
  );
};

const BoxChat = memo(BoxChatComponent);

export default BoxChat;
