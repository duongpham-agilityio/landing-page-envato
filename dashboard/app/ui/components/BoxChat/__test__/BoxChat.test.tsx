import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import BoxChat from '@/ui/components/BoxChat';

describe('BoxChatComponent', () => {
  test('BoxChat component renders correctly', () => {
    const { container } = render(<BoxChat />);
    expect(container).toMatchSnapshot();
  });

  test('calls handleSendMessage when the send button is clicked', async () => {
    render(<BoxChat />);

    const sendButton = screen.getByTestId('btn-send');
    await fireEvent.click(sendButton);

    waitFor(() => {
      expect(sendButton).toBeDisabled();
    });
  }, 2000);

  test('calls handleSendMessage when Enter key is pressed', () => {
    const handleSendMessage = jest.fn();
    render(<BoxChat />);

    const input = screen.getByPlaceholderText('Type your message...');
    fireEvent.change(input, { target: { value: 'Test message' } });

    fireEvent.keyDown(input, { key: 'Enter' });

    waitFor(() => {
      expect(handleSendMessage).toHaveBeenCalledWith('Test message');
    });
  });
});
