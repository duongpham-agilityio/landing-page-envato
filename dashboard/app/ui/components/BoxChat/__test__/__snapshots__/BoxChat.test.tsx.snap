// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxChatComponent BoxChat component renders user have roomChatId 1`] = `
<div>
  <div
    class="css-1pphdyr"
  >
    <div
      class="css-z7702n"
    >
      <h3
        class="chakra-heading css-1rl4kbt"
      >
        team chat
      </h3>
    </div>
    <div
      class="css-1iwi7g1"
    >
      <div
        class="chakra-spinner css-nkpswe"
      >
        <span
          class="css-8b45rq"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`BoxChatComponent BoxChat component renders user without roomChatId 1`] = `
<div>
  <div
    class="css-1pphdyr"
  >
    <div
      class="css-z7702n"
    >
      <h3
        class="chakra-heading css-1rl4kbt"
      >
        team chat
      </h3>
    </div>
    <div
      class="css-1iwi7g1"
    >
      <div
        class="chakra-spinner css-nkpswe"
      >
        <span
          class="css-8b45rq"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
